[VSR] Protocol batching (multi-operation)

batiati opened the pull request:
### VSR Protocol Batching - Multi-operation

This PR enhances the protocol batching proposal introduced in [PR #2617](https://github.com/tigerbeetle/tigerbeetle/pull/2617), following the same guidelines but allowing clients to submit batches containing different `operations` within the same request.

For more details about the motivation behind shifting the responsibility of batching from the state machine (aka `Demuxer`) to the VSR protocol, please refer to [PR #2617](https://github.com/tigerbeetle/tigerbeetle/pull/2617).

#### Motivation

Previously in #2617, we assumed that the batched request payload would always consist of a single operation. However, since the goal of protocol batching is to abstract the payload encoding from the state machine, enabling multi-operation payloads became the natural choice.

The major change lies in how the state machine process multiple cycles of prefetches for each operation, transitioning from `prefetch -> execute -> compact` to `prefetch¹ -> execute¹ -> prefetch² -> execute² -> prefetchⁿ -> executeⁿ -> compact` (where _n_ is the number of batches).

Interestingly, this change enabled batching queries at no additional cost. The primary reason queries were not batchable in #2617 was the hard limit on LSM scans, which could only run a single query during prefetching. By splitting the work into multiple prefetches, multiple queries are now allowed, up to the `limit` of results that can fit into a reply message.

To reduce the cost of context switching between prefetch and execute, adjacent batches of the same operation can, in some cases, be prefetched together in one go (for example, `create_{accounts,transfers}` and `lookup_{accounts,transfers}`), in the following sequence: `prefetch¹⁺²⁺ⁿ -> execute¹ -> execute² -> executeⁿ -> compact` (where _n_ is the number of adjacent batches).

#### VSR Changes

A new reserved VSR operation, `batched`, has been introduced to indicate whether a request is batched. Clients from older versions, which are unaware of this new operation, will always send non-batched requests, ensuring compatibility with the new protocol version.

**How it works:**

- The client (`tb_client` library) is responsible for accumulating concurrent requests, encoding them into batches, and sending them as a single VSR request with the reserved operation `batched`. The underlying state-machine operation for each batch is encoded as part of the batch trailer.

- Upon receiving a batched request, the replica decodes the message body, splits it into multiple batch payloads, and feeds the state machine with independent units of work under the same VSR `op` number.

- Each batch may generate a number of `result`s unrelated to the number of `event`s (e.g., `query_transfers` and `lookup_transfers`). These results are encoded into a batched reply and sent back to the client.

- The client decodes the reply's body into multiple payloads and invokes the respective user callback for each one.

**Encoding layout:**

- Batched requests use a portion at the end of the message body (the batch `trailer`) to encode batch metadata. As a result, batched requests can hold fewer events than non-batched ones.

- The `trailer` size is determined by the number of batches, requiring 4 bytes per batch (a 24-bit integer for the size and an 8-bit integer for the operation), plus a _postamble_ at the very end of the trailer containing the `batch_count`.

- To minimize the number of copies during encoding, the trailer is written from the end of the message towards the beginning. For example, the last element of the array corresponds to the number of events in the first batch (as in a stack).

- Batches of different operations may require padding between them to maintain alignment. Padding bytes are always zeroed. The padding between batches is not stored in the encoding and should be calculated based on the size/alignment expected by the state machine (e.g., by calling `std.mem.alignForward`). The `vsr.batch` implementation abstracts this complexity.

### Example

Consider a **reply** message containing 4 different operations:


| # | operation            | operation code | result size |
|---|----------------------|----------------|-------------|
| 1 |`lookup_accounts`     | 131            | 128 bytes   |
| 2 |`create_accounts`     | 129            |   8 bytes   |
| 3 |`lookup_transfers`    | 132            | 128 bytes   |
| 4 |`get_account_balances`| 134            | 128 bytes   |
-----------------------------------------------------------

```
size == 1076   message.body_used().len == 820 bytes    trailer == 20 bytes
               payload == 800 bytes                    batches (4 bytes each)          postamble (4 bytes)   
┌──────┐ ┌───────────────────────────────────────────┐┌─────────────────────────────────────┐
│      │ │┌─────────┐┌─────────┐┌────────┐┌─────────┐││┌───┐┌───┐┌───┐┌───┐            ┌───┐│
│ VSR  │ ││512 bytes││ 24 bytes││ 0 bytes││264 bytes││││264││  0││ 24││512│← size      │ 4 ││← batch_count
|Header| |│4 results││3 results││0 result││2 results││││134||132││129││131│← operation │ 0 ││← reserved
│      │ │└───▲─────┘└───▲─────┘└────▲───┘└──▲──────┘││└─┬─┘└─┬─┘└─┬─┘└─┬─┘            └───┘│
└──────┘ └────┼──────────┼───────────┼───────┼───────┘└──┼────┼────┼────┼───────────────────┘
              │          │           │       │  8 bytes  │    │    │    │  
              │          │           │       └──padding──┘    │    │    │  
              │          │           │                        │    │    │  
              │          │           └────────────────────────┘    │    │  
              │          │                                         │    │  
              │          └─────────────────────────────────────────┘    │  
              │                                                         │  
              └─────────────────────────────────────────────────────────┘  
```

- The `tb_client` avoids batching user requests if the combined payload length and trailer exceed the maximum message body size. Instead, it sends a non-batched request to preserve maximum capacity (up to 8190 events per message).

- The `tb_client` attempts to pack batches of the same operation sequentially, reducing the need for multiple prefetch cycles.

- The client can pack as many batches as the sum of payloads and trailer allows.

### This PR

See the commit messages for implementation details (most of which are the same commits from PR #2617).

This PR includes VOPR and Vortex (Java driver only) tests. However, Vortex can only process a single operation at a time, so multi-operation batches are not explored.

batiati committed:
[VSR] Add `batch_count` to the message header.
Since protocol batching is integrated into VSR, a protocol-level mechanism is required to indicate whether the message body can be processed as a single unit (`batch_count = 0`) or needs to be decoded into multiple payloads and commited sequentially within the same request (`batch_count > 0`).

Some operations might want to disable protocol batching by always sending `batch_count = 0`.
Clients might choose whether or not to submit batched messages to avoid wasting bytes with the trailer, maximizing the number of events they can pack into a single request.
Additionally, old clients will maintain compatibility since this field will always be zero in versions prior to this feature.

batiati committed:
Add batch.zig

batiati committed:
[StateMachine] Make alignment checks operation-specific.
Accept non-aligned memory and check the alignment required by the specific state machine operation.

Previously, we enforced the state machine signatures to receive aligned slices as `[]align(16) u8` and `*align(16) [message_body_size_max]u8`.

This approach would cause friction with custom state machines by requiring all events/results to be aligned by a constant. It's also already inconsistent with results such as `CreateAccountsResult` and `CreateTransfersResult`, which are 8 bytes long. For example, when replying 3 results, we don't add any padding to keep it `align(16)`,  we would return 24 bytes instead of 32 bytes.

However, the main motivation for this change is protocol batching. When passing chunks of memory to the state machine for reading/writing, those chunks are aligned to the corresponding Event and Result, but not always aligned to 16 (especially when writing results to a buffer of `CreateAccountsResult` and `CreateTransfersResult`).

batiati committed:
[Client] Refactor `tb_client` and remove batching support.
- Completely remove the `DemuxerType`, the old decoder that relied on the `result.index` field to split replies into multiple batches.
- Remove all batching-related fields from `Packet` (some may be needed again soon).
- Refactor `tb_client` to simplify the handling of the submitted queue and remove the `pending` FIFO.
- This refactor essentially restores `tb_client` to its state before batches were introduced, with additional improvements in naming and code organization.

batiati committed:
Implement batching logic
- Introduce `BatchEncoder` and `BatchDecoder` types to write and parse the batch trailer during send/receive operations. With unit tests.
- Replica is aware of batching, while StateMachine is now entirely unaware. Replica is responsible for splitting the message body into batches and feeding the StateMachine during `commit`, as well as writing the batched reply. See the changes required regarding the `prepare_timestamp.
- `tb_client` uses the same `Packet.next` field to link packets within the same batch. `tb_client` groups packets of the same operation and uses the `BatchEncoder` to write the message. During callbacks, the same linked packets are used to notify the caller of each individual batch. Note: This is an unoptimized implementation that may change soon (see the comment on the code).
- The Echo client is now capable of batching events and validating the replies.

batiati committed:
[Client] tb_client batching
- Refactored `tb_client` with improved naming and organization to distinguish between callbacks, functions called by the user thread, and functions called by the IO thread. Also disambiguated names such as `on_submit`,  `submit` and `request`.
- Unified shutdown and eviction handling.
- Reintroduced the `pending` and `submitted` queues, eliminating the O(n) when batching packets.
- Split `Packet` and `Packet.Extern` into two distinct types, hiding implementation details from the end user.
- Added more assertions regarding the packet phase and the expected state of the linked list fields.

batiati committed:
[Node] Remove `BufferSize`.
The type `BufferSize` was a packed struct with two counts, repurposing the `packet.data_size` to hold the size of both events and results. Now it uses `packet.tag`, a new user-defined exposed in `Packet.Extern` (instead of wasting space with padding bytes).

batiati committed:
[Vortex] Update to the new `Packet.Extern` api.

batiati committed:
[aof] Update to the new `batch_count` api

batiati committed:
[VSR] Support batched input validation
Validates the batch format instead of panicking.

batiati committed:
[VOPR] Add support for batched requests.

batiati committed:
Fix tests

batiati committed:
Verify the batch format after encoding.

batiati committed:
[Vortex] Add support for async operations to the Java driver.
It allows testing batched requests.

batiati committed:
Review

batiati committed:
Fix cancel inflight request

batiati committed:
Batch: remove `can_add`.

batiati committed:
[VSR] Remove the `batch_count` header and introduce a new reserved operation `batched` to indicate a batched payload.

batiati committed:
[VSR] Batch encoder and decoder refactor.

The batching format now supports multiple different operations within the same message, requiring a more comprehensive trailer containing the `operation` and `size`, along with padding and alignment safeguards.

The encoder/decoder manages each operation's size and alignment through functions implemented by the caller. The caller, being aware of the StateMachine and variations according to the client release, ensures compatibility.

Currently, only basic unit tests have been implemented. Edge case testing is still required.

batiati committed:
[VSR] Multi-operation protocol batching.

batiati committed:
[Vopr] Support multi-operation batching

batiati committed:
[Clients] Update bindings

batiati committed:
[Clients] Test multi-operation concurrent requests

batiati committed:
[AOF] Fix

batiati committed:
[Vopr] Fix requests_replied increment.

batiati committed:
Simplify the commit state machine:
- Remove the additional `prefetch_batched` commit stage.
- Call `commit_reply_setup` only after all batches have been prefetched.

batiati committed:
[AOF] Allow replay batched requests

batiati committed:
Fix testinig state machine.

batiati committed:
Replica: op max safety counter

batiati commented on src/state_machine.zig:
```diff
@@ -709,7 +702,7 @@ pub fn StateMachineType(
             client_release: vsr.Release,
             operation: Operation,
-            input: []align(16) const u8,
+            input: []const u8,
```
Vopr seed: `./zig/zig build -Drelease vopr -- --lite 12293026746782977520 `

Cause: Vopr workload generating too many batched events.

Root cause: `StateMachine.input_valid` is not enough to validate a batched request. It validates each batch isolated, but we must also validate all batches combined and check if they fit into the commit limits.

I :heart: VOPR!

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+const constants = @import("../constants.zig");
+const vsr = @import("../vsr.zig");
+
+const Postamble = packed struct(u32) {
```
Why `packed` instead of `extern`?

batiati replied:
Just to be symmetrical with `TrailerItem`.

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+            current_payload_size: usize,
+            current_batch_count: u16,
+            next_operation: vsr.Operation,
+            next_payload_size: usize,
```
Naming wise, are `current_payload_size` and `next_payload_size` using "payload" to refer to two different things?

The former is the cumulative size of the batch bodies. The latter is a single batch body.

What about:

```suggestion
            current_payload_size: usize,
            current_batch_count: u16,
            next_batch_operation: vsr.Operation,
            next_batch_size: usize,
```

:thinking: ...Though, that isn't great either. There are two different levels of batches -- the data we are encoding is a batch (of batches), and each chunk corresponding to an operation is itself a batch.

Can we disambiguate? What if we just (consistently) used a different synonym for the former? e.g. "parcel" / "protocol parcel" / `Operation.parcel` / `vsr/parcel.zig`?

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+                    alignment,
+                ));
+                break :padding padding;
+            };
```
Refactor so that the asserts cover both branches:

```suggestion
            const padding: u16 = padding: {
                if (options.next_payload_size == 0) {
                    break :padding 0;
                } else {
                    break :padding align_padding(
                        options.current_payload_size,
                        alignment,
                    );
                }
            };
            // Assuming aligned buffers, so the first batch will not require padding.
            assert(padding == 0 or options.current_payload_size > 0);
            assert(std.mem.isAligned(
                options.current_payload_size + padding,
                alignment,
            ));
```

sentientwaffle replied:
The 256-byte message header is also part of the message -- should we assert that we are aligned relative to that too?

sentientwaffle commented on src/vsr/batch.zig:
Per tigerstyle, avoid `usize` -- in some of the cases in this file (e.g. `align_padding()`), using `usize` might lead to unexpected differences when building on wasm with 32-bit.

Likewise for how we use pointer alignment, though that might be trickier. e.g. image a 32-bit client talking to a 64-bit replica.

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+            if (body.len < trailer_size) return error.BatchInvalid;
+
+            const payload_end = body.len - trailer_size;
+            const payload: []const u8 = body[0..payload_end];
```
More verbose, but better to be explicit:
```suggestion
            const payload_with_padding_size = body.len - trailer_size;
            const payload_with_padding: []const u8 = body[0..payload_with_padding_size];
```
(I am having trouble keeping track of which counts include padding :sweat_smile:)

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+const constants = @import("../constants.zig");
+const vsr = @import("../vsr.zig");
+
+const Postamble = packed struct(u32) {
```
Checking my understanding -- is the reason that the `Postamble` is at the end of the body rather than the beginning (since it is fixed size, unlike the list of trailer items) to avoid adding padding afterwards?

batiati replied:
Yes, since it's 4 bytes, payloads with `align(8)` and `align(16)` would require padding just at the beginning of the message.

Also, it makes sense as the entire trailer is written in the reverse order.

sentientwaffle commented on src/vsr/batch.zig:
A module-level comment with some terminology definitions might be useful -- e.g. "element", "payload", "trailer", "trailer item", "postamble", "batch".

Also, maybe a simplified version of your diagram, to show the schema? In particular, pointing out the different alignments/paddings -- I find those hard to keep track of.

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+            return aligned[0..size];
+        }
+
+        /// Records how many bytes were writen in the slice previously acquired by `writable()`.
```
```suggestion
        /// Records how many bytes were written in the slice previously acquired by `writable()`.
```

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+            if (elements_size_total + trailer_padding != payload.len) {
+                return error.BatchInvalid;
+            }
+            if (trailer_padding > 0 and !stdx.zeroed(payload[payload.len - trailer_padding ..])) {
```
Simplify to
```suggestion
            if (!stdx.zeroed(payload[payload.len - trailer_padding ..])) {
```

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+    /// at the beginning of the batch to maintain alignment.
+    /// Use `std.mem.alignForward` to compute the correct starting position for the required
+    /// alignment. Padding bytes are zeroed.
+    size: u24,
```
Maybe clarify in the comment that `size` doesn't include the padding?

batiati replied:
But it does include paddind!
We really need comments here :grin:

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+/// The trailer is a `[]BatchItem` containing the operation and the number of elements
+/// in each batch. To encode the trailer, `batch_count * @sizeOf(BatchItem)` bytes are
+/// needed, plus a `Postamble`.
+fn trailer_total_size(batch_count: u16) usize {
```
It is confusing that the "total trailer size" _doesn't_ include padding, while the `trailer_size` (returned by `total_size`) _does_ include padding. Maybe `trailer_unpadded_size`?

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+    return (@as(usize, batch_count) * @sizeOf(TrailerItem)) + @sizeOf(Postamble);
+}
+
+fn align_padding(target: anytype, alignment: usize) u16 {
```
:thinking: Maybe `padding_for_alignment` or `alignment_padding`? `align_padding` makes it sound like we are aligning the _padding_, which isn't the case.

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+                trailer[0 .. trailer.len - @sizeOf(Postamble)],
+            ));
+            assert(trailer_items.len == self.batch_count);
+            if (constants.verify) {
```
Add an explanation that these asserts are paired with corresponding setup in `.writable()` -- without that context they don't make sense.

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+
+            const buffer: []u8 = self.buffer.?;
+            const padding: u16 = align_padding(
+                buffer[self.buffer_index..],
```
Passing a pointer to `align_padding` and using the result to inform our data layout makes me nervous -- could the result vary if by chance `buffer` was allocated on a particular offset? Could the result vary depending on whether we are a 32/64-bit build?

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+            assert(padding == 0 or self.buffer_index > 0);
+
+            if (buffer.len <= self.buffer_index + padding + trailer_size) {
+                // Insuficient space for one more batch.
```
```suggestion
                // Insufficient space for one more batch.
```

sentientwaffle commented on src/state_machine/workload.zig:
```diff
@@ -303,124 +402,436 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                                    Event,
+                                    action_comptime,
+                                    writable,
+                                ) orelse break; // Insuficient space.
```
```suggestion
                                ) orelse break; // Insufficient space.
```

sentientwaffle commented on src/state_machine/workload.zig:
```diff
@@ -303,124 +402,436 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                                    Event,
+                                    action_comptime,
+                                    writable,
+                                ) orelse break; // Insuficient space.
```
```suggestion
                                ) orelse break; // Insufficient space.
```

sentientwaffle commented on src/vsr/batch.zig:
```diff
@@ -0,0 +1,804 @@
+        }
+    };
+
+    fn run(options: struct {
```
Rewrote this function here since I had too many suggestions and wanted to experiment: https://github.com/tigerbeetle/tigerbeetle/commit/6d6492c2e7c6ab3f044e61b0fb33feb2ca62489d

Primary goal is to not rely on `BatchEncoder` when generating the data to be encoded. We are testing encoder, so we shouldn't trust it!

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -21,13 +21,12 @@ pub fn ClientType(
             pub const Callback = *const fn (
                 user_data: u128,
-                operation: StateMachine.Operation,
+                operation: vsr.Operation,
```
~Is the reason for `s/StateMachine/vsr` that using the state machine operations at the "top level" is essentially deprecated?~ Ah right, it is because we need to be able to pass `Operation.batched`.

batiati committed:
Fix batch size

- The maximum number of entries for prefetching should account for lookups, not just the create operation.  The input size is constrained by the runtime parameter `batch_size_limit`, while the reply is only limited by `message_body_size_max`. This can result in the maximum events per batch for `lookup_{accounts,transfers}` being greater than for `create_{accounts,transfers}`.

- Rename `groove_options.batch_value_count_max` to `value_count_max` to clarify the distinction between the number of entries and prefetch.

- In `batch.writable()`, treat "no space for one more batch" as a `null` slice, while an empty slice indicates no space left for an additional element (though a zeroed batch would be allowed).

- Improve the unit test for `batch`.

batiati committed:
Review

- Addressed the `usize` type and the issue passing a pointer for calculating the trailer alignment. Now, we use just an index as `u32`, assuming index zero is always aligned.
  Also replaced `usize` with `u32` in several places, including the state machine (e.g., commit).

- Improved naming and documentation.

- Fixed Replica validation for batched requests in `input_valid`, ensuring the reply size is validated and the total batched input is asserted against `batch_size_max`.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -56,6 +58,131 @@ pub fn ContextType(
+            }.alignment,
+        });
+
+        fn result_expected_size(
```
Should this be `reply_body_expected_size`? So that it isn't confused with `StateMachine.Result`.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -240,6 +360,10 @@ pub fn ContextType(
             self.signal.stop();
             self.thread.join();
+
+            assert(self.submitted.pop() == null);
+            assert(self.pending.pop() == null);
```
Or:
```suggestion
            assert(self.submitted.empty());
            assert(self.pending.empty());
```

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
+                    .next_batch_operation = vsr.Operation.from(StateMachine, operation),
+                    .next_batch_size = packet.data_size,
+                });
+                if (self.batch_size_limit.? <
+                    encoded_events.payload_size + encoded_events.trailer_size) continue;
+
+                const encoded_results = BatchDecoder.encoded_total_size(.{
+                    .context = {},
+                    .current_payload_size = root.batch_results_size,
+                    .current_batch_count = @max(1, root.batch_count) + 1,
+                    .next_batch_operation = vsr.Operation.from(StateMachine, operation),
+                    .next_batch_size = results_size,
+                });
+                // While the request size is constrained by `batch_size_limit`,
+                // the reply size is only limited by `message_body_size_max`.
+                if (encoded_results.payload_size + encoded_results.trailer_size >
+                    constants.message_body_size_max) continue;
```
These two conditions are nearly the same. Matching the order makes their similarities more obvious.

```suggestion
                if (encoded_events.payload_size + encoded_events.trailer_size >
                    self.batch_size_limit.?) continue;

                const encoded_results = BatchDecoder.encoded_total_size(.{
                    .context = {},
                    .current_payload_size = root.batch_results_size,
                    .current_batch_count = @max(1, root.batch_count) + 1,
                    .next_batch_operation = vsr.Operation.from(StateMachine, operation),
                    .next_batch_size = results_size,
                });
                // While the request size is constrained by `batch_size_limit`,
                // the reply size is only limited by `message_body_size_max`.
                if (encoded_results.payload_size + encoded_results.trailer_size >
                    constants.message_body_size_max) continue;
```

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
+
+                if (root.batch_count == 0) {
+                    assert(root.batch_next == null);
+                    // Although the protocol allows `batch_count == 1`,
+                    // the client never packages a single packet into a batched request.
+                    // Instead, it sends a non-batched request to avoid wasting message
+                    // body bytes with the batch trailer.
```
Possibly worth discussing in the W+T, but it might make sense to just always send batched requests, to reduce client dimensionality.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
-                    },
-                    else => false,
-                };
+        fn batch_merge(self: *Context) ?*Packet {
```
Why do we need to perform batching both here _and_ in `packet_enqueue`?

batiati replied:
We might want to improve this algorithm, `packet_enqueue` joins batches of the same operation while `batch_merge` joins multioperations. It keeps operations together without having to find where to insert between two linked list nodes.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
-        fn submit(self: *Context, packet: *Packet) void {
+        /// Sends the packet (the entire batched linked list of packets) through the vsr client.
+        /// Always called by the io thread.
+        fn packet_send(self: *Context, packet: *Packet) void {
```
Maybe call this `root` (as we do elsewhere) to hint that this is the beginning of a chain of packets?

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
+            assert(self.batch_size_limit == null);
+            assert(result.batch_size_limit > 0);
+
+            // The client might have a smaller message size limit.
+            maybe(constants.message_body_size_max < result.batch_size_limit);
+            self.batch_size_limit = @min(result.batch_size_limit, constants.message_body_size_max);
```
When is this true? During a test?

batiati replied:
Clients should be compatible if new versions advertise a larger message limit (e.g. `message_body_size_max` > 1MiB), or if we want to support running a debug client against a release server.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
+            self.cancel_queued_packets();
+        }
+
+        fn client_result_callback(
```
I know this was previously called `on_result`, but "result" typically refers to `StateMachine.Result`. Reply is more accurate:

```suggestion
        fn client_reply_callback(
```

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
+                assert(packet.batch_count > 1);
+
+                var decoder = BatchDecoder.init({}, reply) catch {
+                    std.debug.panic("client_result_callback: invalid batch: " ++
+                        "reply.len={}", .{
+                        reply.len,
+                    });
```
Maybe log the `timestamp` and `packet.batch_count` as well, since we have those handy?

This would be a really bad error for a client to hit -- it _should_ be unreachable, but just in case, we should include anything which could possibly be used to troubleshoot.

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
+                };
+                assert(packet.batch_count == decoder.batch_count());
+
+                var it: ?*Packet = packet;
+                while (it) |batched| {
+                    it = batched.batch_next;
```
```suggestion
                var packet_iterator: ?*Packet = packet;
                while (packet_iterator) |batched| {
                    packet_iterator = batched.batch_next;
```

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -290,317 +391,499 @@ pub fn ContextType(
         ) void {
             const user_data: UserData = @bitCast(raw_user_data);
             const self = user_data.self;
             const packet = user_data.packet;
```
Or maybe `root`?

sentientwaffle commented on src/clients/c/tb_client/packet.zig:
```diff
@@ -55,11 +56,121 @@ pub const Packet = extern struct {
         }
     };
+
+    const Phase = enum(u8) {
```
Should this be [`extern enum`](https://ziglang.org/documentation/master/#extern-enum)?

> By default, enums are not guaranteed to be compatible with the C ABI

(Or maybe that sentence doesn't apply when the enum integer type is explicitly declared.)

batiati replied:
It's not part of the `Packet.Extern` structure, so never exposed to the C ABI.

sentientwaffle commented on src/clients/c/tb_client/packet.zig:
```diff
@@ -55,11 +56,121 @@ pub const Packet = extern struct {
+            },
+            .complete => {
+                // The packet pointer isn't available afer completed,
+                // it may be dealocated by the user;
```
```suggestion
                // it may be deallocated by the user;
```

sentientwaffle commented on src/clients/c/tb_client/signal.zig:
```diff
@@ -86,7 +87,7 @@ pub const Signal = struct {
 
     fn wait(self: *Signal) void {
-        assert(!self.stop_requested());
+        maybe(self.stop_requested()); // Stop requested while processing the event.
```
Does this comment mean that the stop was requested during `on_signal_fn()`?

batiati replied:
You independently found the bug fixed at #2689 :smile:

sentientwaffle commented on src/clients/c/test.zig:
```diff
@@ -152,6 +157,33 @@ test "c_client echo" {
         var completion = Completion{ .pending = concurrency_max };
 
+        const operation = operations[
+            prng.random().intRangeAtMost(
+                usize,
+                0,
+                operations.len - 1,
+            )
```
```suggestion
            prng.random().uintAtMost(
                usize,
                operations.len - 1,
            )
```

sentientwaffle commented on src/clients/java/src/test/java/com/tigerbeetle/IntegrationTest.java:
```diff
@@ -1306,32 +1305,47 @@ public void testAsyncTasks() throws Throwable {
-            for (int i = 0; i < TASKS_COUNT; i++) {
+            for (int i = 0; i < TASKS_COUNT; i += 1) {
 
                 final var transfers = new TransferBatch(1);
                 transfers.add();
```
Can this move into the below `else`  statement?

sentientwaffle commented on src/clients/java/src/test/java/com/tigerbeetle/IntegrationTest.java:
```diff
@@ -2369,7 +2385,7 @@ public Server(final String label) throws IOException, Exception, InterruptedExce
             }
 
             this.process = new ProcessBuilder()
-                    .command(new String[] {exe, "start", "--addresses=0", "--development", tb_file})
```
Why remove `--development`?

batiati replied:
I found divergences between Java and Go integrated tests, mainly because Go runs without the `--development` flag.

Anyway, it was an accidental commit, I intend to drop this flag for all integration tests in another PR.

sentientwaffle commented on src/vsr.zig:
```diff
@@ -266,6 +267,8 @@ pub const Operation = enum(u8) {
     /// The value 5 is is reserved for release-upgrade requests.
     upgrade = 5,
+    /// Batched message that includes multiple state machine operations. See `vsr.batch`.
+    batched = constants.vsr_operations_reserved - 1,
```
Why not `= 6`?

sentientwaffle commented on src/testing/vortex/java_driver/src/main/java/Main.java:
```diff
@@ -55,24 +68,42 @@ record Driver(Client client, Reader reader, Writer writer) {
    * stdout.
-   */
-  void next() throws IOException {
+      * @throws ExecutionException
+      * @throws InterruptedException
+      */
```
```suggestion
   * @throws ExecutionException
   * @throws InterruptedException
   */
```

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -124,75 +124,13 @@ pub fn StateMachineType(
-
-        const batch_value_count_max = batch_value_counts_limit(config.message_body_size_max);
-
+        const tree_values_count_limit = tree_values_count_max(config.message_body_size_max);
```
Why "limit" rather than "max"? Typically we use "limit" to indicate a runtime upper bound, vs "max" for a comptime upper bound.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -709,59 +771,83 @@ pub fn StateMachineType(
+                    if (input.len > 0 and
+                        !std.mem.isAligned(@intFromPtr(input.ptr), @alignOf(Event)))
+                    {
+                        log.info("Invalid input alignment for {s}", .{
+                            @tagName(operation),
+                        });
```
This seems like it should probably be `log.err` or at least `log.warn`, since (iiuc) it would indicate either a client or a replica bug. (And in case of the former, we should include the client release in the log message.)

Also, please include the replica index prefix (via `self.forest.grid.superblock.replica_index`).

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -709,59 +771,83 @@ pub fn StateMachineType(
-        /// Updates `prepare_timestamp` to the highest timestamp of the response.
-        pub fn prepare(
-            self: *StateMachine,
+        /// Returns the logical time increment (in nanoseconds) for the highest
+        /// timestamp of the response.
```
```suggestion
        /// Returns the number of nanoseconds that this prepare will consume.
```

The original wording was confusing -- it sounds like a mix of "this is a duration" and "this is an absolute timestamp". Not sure if my version is an improvement. :sweat_smile:

sentientwaffle reviewed (commented):
WIP

batiati commented:
Ref: #2787
